### PR TITLE
C++: Treat implicit end of body of non`-void` function as `Unreached`

### PR DESCRIPTION
**Pull request by dbartol**
When the extractor can't prove that control flow will never reach the end of a non-`void`-returning function without reaching an explicit `return` statement, it inserts an implicit `return` without an operand. If control actually reaches this point, the behavior is undefined.

We were previously generating invalid IR for these implicit `return` statements, because the lack of an operand meant that there was no definition of the return value variable along that path. Instead, I've changed the IR generation to emit an `Unreached` instruction for the implicit `return`. This ensures that we don't create a control flow edge from the end of the body to the function epilogue.

The change to the range analysis test avoids having that test depend on the previous bad IR behavior, while still preserving the original spirit of the test.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 1]
```diff
@@ -147,8 +150,14 @@ class TranslatedReturnValueStmt extends TranslatedReturnStmt, TranslatedVariable
...
 }
 
+/**
+ * The IR translation of a `return` statement that does not return a value. This includes implicit
+ * return statements at the end of `void`-returning functions.
+ */
 class TranslatedReturnVoidStmt extends TranslatedReturnStmt {
-  TranslatedReturnVoidStmt() { not stmt.hasExpr() }
+  TranslatedReturnVoidStmt() {
+    not stmt.hasExpr() and not hasReturnValue(stmt.getEnclosingFunction())
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:159]: A return statement in a function with `void` as its return type can have a `void`-typed expression - what happens then?

Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:159]: Great question. I’ll add a test case (and probably a fix).

Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll:194]: Fix and test case added. We didn't handle this correctly before, either. Good catch.

**Comment by rdmarsh2** [comment]
oh, missed the failing tests - looks like some noise in block numbering in the syntax zoo consistency failures and a real change in the IR value numbering test

**Comment by dbartol** [comment]
A couple of the failures in `syntax-zoo` were actually fixed correctly by this change, but the rest of that directory was noise.

Also, whoever wrote the original value number tests had something against `return` statements:) I just made several functions return `void`, since they weren't returning a value anyway.

**Pull request merged**